[Merged by Bors] - feat(Analysis/Convex/Side): affine combinations and sides of faces of a simplex

jsm28 opened the pull request:
Add lemmas saying when (in terms of the weights in the combinations) two affine combinations of the vertices of a simplex lie (strictly or weakly) on the same or opposite sides of a face of that simplex opposite a vertex.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

When merging, all the commits will be squashed into a single commit
listing all co-authors.

Co-authors in the squash commit are gathered from two sources:

First, all authors of commits to this PR branch are included. Thus,
one way to add co-authors is to include at least one commit authored by
each co-author among the commits in the pull request. If necessary, you
may create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

Second, co-authors can also be listed in lines at the very bottom of
the commit message (that is, directly before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines
at the bottom of the commit message (before the `---`, and also before
any "Co-authored-by" lines) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [x] depends on: #xyz [optional extra text]

-->

- [x] depends on: #31204
- [x] depends on: #32150
- [x] depends on: #32161
- [x] depends on: #32164

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

jsm28 committed:
feat(LinearAlgebra/AffineSpace/Combination): `affineCombination_mem_affineSpan_image`

Add a lemma giving a condition for an affine combination to lie in the
affine span of a subset of the indexed family of points (the reverse
direction of `eq_affineCombination_of_mem_affineSpan_image`, which is
the only current `image` lemma in this file, roughly).

jsm28 committed:
feat(Analysis/Convex/Side): affine combinations and sides of faces of a simplex

Add lemmas saying when (in terms of the weights in the combinations)
two affine combinations of the vertices of a simplex lie (strictly or
weakly) on the same or opposite sides of a face of that simplex
opposite a vertex.

jsm28 committed:
Merge branch 'master' into simplex_side_affinecombination

jcommelin commented on Mathlib/Analysis/Convex/Side.lean:
```diff
@@ -829,3 +829,219 @@ theorem isPreconnected_setOf_sOppSide (s : AffineSubspace ℝ P) (x : P) :
+      simp only [lineMap_apply, vsub_eq_sub, Pi.vadd_apply', Pi.smul_apply, Pi.sub_apply,
+        smul_eq_mul, vadd_eq_add] at hs₂
+      rw [← neg_sub (w₁ i) (w₂ i), mul_neg, div_mul_cancel₀ _ hp.ne'] at hs₂
+      simp at hs₂
```
There seems to be quite a lot of duplication here. Are we missing a useful lemma/abstraction?

jcommelin commented on Mathlib/Analysis/Convex/Side.lean:
```diff
@@ -829,3 +829,219 @@ theorem isPreconnected_setOf_sOppSide (s : AffineSubspace ℝ P) (x : P) :
+    intro h0'
+    apply h.right_notMem
+    rw [range_faceOpposite_points]
+    exact affineCombination_mem_affineSpan_image hw₂ (by simp [h0']) s.points
```
Idem dito.

jsm28 replied:
I've opened #32150 to add a lemma `affineCombination_mem_affineSpan_faceOpposite_iff`, made this PR depend on that one and refactored all the proofs here that used `range_faceOpposite_points` to use that new lemma. (The remaining case of duplication you identified will require further definitions and lemmas elsewhere to resolve.)

jcommelin reviewed (commented):
I flagged a few places where there seems to be duplication. I wonder if you have ideas to abstract these out.

jsm28 committed:
feat(LinearAlgebra/AffineSpace/Simplex/Basic): `affineCombination_mem_affineSpan_faceOpposite_iff`

Add a convenience lemma that reflects common patterns appearing in #31205 and pointed out in review of that PR.

```lean
lemma affineCombination_mem_affineSpan_faceOpposite_iff {n : ℕ} [NeZero n] {s : Simplex k P n}
    {w : Fin (n + 1) → k} (hw : ∑ i, w i = 1) {i : Fin (n + 1)} :
    Finset.univ.affineCombination k s.points w ∈
      affineSpan k (Set.range (s.faceOpposite i).points) ↔ w i = 0 := by
```

jsm28 committed:
Merge branch 'affineCombination_mem_affineSpan_faceOpposite_iff' into simplex_side_affinecombination

jsm28 committed:
Refactor using `affineCombination_mem_affineSpan_faceOpposite_iff`

jsm28 committed:
Remove stray space

jsm28 committed:
feat(LinearAlgebra/AffineSpace/Basis): spaces of coordinates

Define `fintypeAffineCoords`, the `AffineSubspace` of `ι → k` (for
`Fintype ι`) where coordinates sum to `1`, and the analogous
`finsuppAffineCoords` for a general `ι`. The `Finsupp` case was
recently discussed in #30854 and is the subject of a TODO (not
addressed) in this file. The `Fintype` case is intended for use in #31205
to refactor a proof based on review there.

jsm28 committed:
feat(Analysis/Convex/Between): `InjOn` lemmas

Add lemmas `Set.InjOn.wbtw_map_iff` and `Set.InjOn.sbtw_map_iff`, analogous to those already present for `Function.Injective`. For use together with #32161 in #31205.

jsm28 committed:
Merge branch 'fintypeAffineCoords' into simplex_side_affinecombination

jsm28 committed:
Merge branch 'injon_btw_map_iff' into simplex_side_affinecombination

jsm28 committed:
Refactor duplicative code using `Set.InjOn.sbtw_map_iff`

jsm28 committed:
Merge branch 'master' into simplex_side_affinecombination

YaelDillies reviewed (approved):
Thanks!

maintainer delegate

jsm28 committed:
Update Mathlib/Analysis/Convex/Side.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

jsm28 committed:
Update Mathlib/Analysis/Convex/Side.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

jsm28 committed:
Merge branch 'master' into simplex_side_affinecombination

jsm28 committed:
Also use `end Affine.Simplex`

jcommelin reviewed (commented):
Thanks :tada:

bors merge
